File: app/src/main/java/com/github/malitsplus/shizurunotes/data/Chara.java
package com.github.malitsplus.shizurunotes.data;

import com.github.malitsplus.shizurunotes.R;
import com.github.malitsplus.shizurunotes.common.I18N;
import com.github.malitsplus.shizurunotes.data.action.PassiveAction;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Chara {

    //basic params
    public int unitId;
    public String unitName;
    public int prefabId;
    public int searchAreaWidth;
    public int atkType;

    public int moveSpeed;
    public double normalAtkCastTime;
    public String actualName;
    public String age;
    public int guildId;

    public String guild;
    public String race;
    public String height;
    public String weight;
    public String birthMonth;

    public String birthDay;
    public String bloodType;
    public String favorite;
    public String voice;
    public String catchCopy;

    public String comment;
    public String selfText;
    public LocalDateTime startTime;
    public String kana;

    public int charaId;
    public String iconUrl;
    public String imageUrl;
    public String position;
    public int positionIcon;
    public String sortValue;

    //preload params
    public int maxCharaLevel;
    public int maxCharaRank;
    public int maxUniqueEquipmentLevel;

    public Property charaProperty;
    public Property rarityProperty;
    public Property rarityPropertyGrowth;
    public Property storyProperty;
    public Property promotionStatus;
    public List<Equipment> equipments;
    public Equipment uniqueEquipment;

    public String getBirthDate(){
        return birthMonth + I18N.getString(R.string.text_month) + birthDay + I18N.getString(R.string.text_day);
    }

    public void setCharaProperty() {
        charaProperty = new Property();
        charaProperty
                .plusEqual(rarityProperty)
                .plusEqual(getRarityGrowthProperty())
                .plusEqual(storyProperty)
                .plusEqual(promotionStatus)
                .plusEqual(getAllEquipmentProperty())
                .plusEqual(getUniqueEquipmentProperty())
                .plusEqual(getPassiveSkillProperty());
    }

    public Property getRarityGrowthProperty(){
        return rarityPropertyGrowth.multiply(maxCharaLevel + maxCharaRank);
    }
    public Property getAllEquipmentProperty(){
        Property property = new Property();
        for (Equipment equipment : equipments){
            property.plusEqual(equipment.getCeiledProperty());
        }
        return property;
    }
    public Property getUniqueEquipmentProperty(){
        Property property = new Property();
        if(uniqueEquipment != null) {
            property
                .plusEqual(uniqueEquipment.getEquipmentData())
                .plusEqual(uniqueEquipment.getEquipmentEnhanceRate().multiply(maxUniqueEquipmentLevel - 1));
        }
        return property;
    }
    public Property getPassiveSkillProperty(){
        Property property = new Property();
        for (Skill skill: skills){
            if (skill.skillClass == Skill.SkillClass.EX1_EVO){
                for (Skill.Action action: skill.actions){
                    if (action.parameter instanceof PassiveAction)
                        property.plusEqual(((PassiveAction)action.parameter).propertyItem(maxCharaLevel));
                }
            }
        }
        return property;
    }

    public List<AttackPattern> attackPatternList = new ArrayList<>();

    public List<Skill> skills = new ArrayList<>();
    //public Map<Skill.SkillClass, Skill> skillMap = new HashMap<>();

}

File: app/src/main/java/com/github/malitsplus/shizurunotes/data/Chara.kt
package com.github.malitsplus.shizurunotes.data

import android.text.format.DateFormat
import androidx.annotation.DrawableRes
import com.github.malitsplus.shizurunotes.R
import com.github.malitsplus.shizurunotes.common.I18N
import com.github.malitsplus.shizurunotes.data.action.PassiveAction
import com.github.malitsplus.shizurunotes.user.UserSettings
import java.text.SimpleDateFormat
import java.time.LocalDateTime
import java.util.*

class Chara: Cloneable {

    @Throws(CloneNotSupportedException::class)
    override fun clone(): Chara {
        return super.clone() as Chara
    }

    fun shallowCopy(): Chara {
        return clone()
    }

    var unitId: Int = 0
    var charaId: Int = 0
    var prefabId: Int = 0
    var searchAreaWidth: Int = 0
    var atkType: Int = 0
    var moveSpeed: Int = 0
    var guildId: Int = 0
    var normalAtkCastTime: Double = 0.0
    @DrawableRes var positionIcon: Int = 0
    var maxCharaLevel: Int = 0
    var maxCharaRank: Int = 0
    var maxUniqueEquipmentLevel: Int = 0
    var maxRarity: Int = 5
    var rarity: Int = 5

    lateinit var actualName: String
    lateinit var age: String
    lateinit var unitName: String
    lateinit var guild: String
    lateinit var race: String
    lateinit var height: String
    lateinit var weight: String
    lateinit var birthMonth: String
    lateinit var birthDay: String
    lateinit var bloodType: String
    lateinit var favorite: String
    lateinit var voice: String
    lateinit var kana: String
    lateinit var catchCopy: String
    lateinit var iconUrl: String
    lateinit var imageUrl: String
    lateinit var position: String
    var comment: String? = null
    var selfText: String? = null
    var sortValue: String? = null

    lateinit var startTime: LocalDateTime

    lateinit var charaProperty: Property
    val rarityProperty = mutableMapOf<Int, Property>()
    val rarityPropertyGrowth = mutableMapOf<Int, Property>()
    lateinit var storyProperty: Property
    lateinit var promotionStatus: Map<Int, Property>
    lateinit var rankEquipments: Map<Int, List<Equipment>>
    var uniqueEquipment: Equipment? = null

    var attackPatternList = mutableListOf<AttackPattern>()
    var skills = mutableListOf<Skill>()

    val birthDate: String by lazy {
        if (birthMonth.contains("?") || birthDay.contains("?")) {
            birthMonth + I18N.getString(R.string.text_month) + birthDay + I18N.getString(R.string.text_day)
        } else {
            val calendar = Calendar.getInstance()
            calendar.set(calendar.get(Calendar.YEAR), birthMonth.toInt() - 1, birthDay.toInt())
            val locale =  Locale(UserSettings.get().getLanguage())
            val format = DateFormat.getBestDateTimePattern(locale, "d MMM")
            SimpleDateFormat(format, locale).format(calendar.time)
        }
    }

    fun setCharaProperty(
        rarity: Int = maxRarity,
        rank: Int = maxCharaRank,
        hasUnique: Boolean = true
    ) {
        charaProperty = Property().apply {
            plusEqual(rarityProperty[rarity])
            plusEqual(getRarityGrowthProperty(rarity, rank))
            plusEqual(storyProperty)
            plusEqual(promotionStatus[rank])
            plusEqual(getAllEquipmentProperty(rank))
            plusEqual(getPassiveSkillProperty(rarity))
            if (hasUnique) {
                plusEqual(uniqueEquipmentProperty)
            }
        }
    }

    fun getSpecificCharaProperty(
        rarity: Int = maxRarity,
        rank: Int = maxCharaRank,
        hasUnique: Boolean = true
    ): Property {
        return Property().apply {
            plusEqual(rarityProperty[rarity])
            plusEqual(getRarityGrowthProperty(rarity, rank))
            plusEqual(storyProperty)
            plusEqual(promotionStatus[rank])
            plusEqual(getAllEquipmentProperty(rank))
            plusEqual(getPassiveSkillProperty(rarity))
            if (hasUnique) {
                plusEqual(uniqueEquipmentProperty)
            }
        }
    }

    private fun getRarityGrowthProperty(rarity: Int, rank: Int): Property{
        return rarityPropertyGrowth[rarity]?.multiply(maxCharaLevel.toDouble() + rank) ?: Property()
    }

    fun getAllEquipmentProperty(rank: Int): Property {
        val property = Property()
        rankEquipments[rank]?.forEach {
            property.plusEqual(it.getCeiledProperty())
        }
        return property
    }

    val uniqueEquipmentProperty: Property
        get() {
            return uniqueEquipment?.getCeiledProperty() ?: Property()
        }

    fun getPassiveSkillProperty(rarity: Int): Property {
        val property = Property()
        skills.forEach { skill ->
            if (rarity >= 5 && skill.skillClass == Skill.SkillClass.EX1_EVO) {
                skill.actions.forEach {
                    if (it.parameter is PassiveAction)
                        property.plusEqual((it.parameter as PassiveAction).propertyItem(maxCharaLevel))
                }
            } else if (rarity < 5 && skill.skillClass == Skill.SkillClass.EX1) {
                skill.actions.forEach {
                    if (it.parameter is PassiveAction)
                        property.plusEqual((it.parameter as PassiveAction).propertyItem(maxCharaLevel))
                }
            }
        }
        return property
    }
}